Clean up SPIFFE-ID and client_id relations (#44)
arndt-s/oauth-spiffe-client-authentication: draft-ietf-oauth-spiffe-client-auth.md

Based on {{RFC7523}} the following request parameters MUST be present to perform client authentication in the context of this specification:

* client_id: MUST identify the client. The `client_id` MAY be the SPIFFE ID of the client, but it MAY also be any other client identifier recognized by the authorization server, such as a URL pointing to a Client ID Metadata Document ({{I-D.ietf-oauth-client-id-metadata-document}}). In either case, the SPIFFE ID carried in the `sub` claim of the JWT-SVID MUST be authorized to authenticate as the client identified by the `client_id`, as described below.
* client_assertion_type: MUST be set to `urn:ietf:params:oauth:client-assertion-type:jwt-spiffe`.
* client_assertion: MUST be a single SPIFFE JWT-SVID.

> Requiring `client_id` makes the client identity explicit in the request, rather than leaving it to be derived from the `sub` claim of the JWT-SVID. It also allows a single SPIFFE ID to be associated with more than one registered client.

Reviewer: Perhaps the "single SPIFFE ID with multiple clients" and "single client_id with multiple SPIFFE IDs should probably get its own section or be discussed in security considerations or as implementation options rather than a single sentence here? 

```suggestion
Requiring `client_id` makes the client identity explicit in the request, rather than leaving it to be derived from the `sub` claim of the JWT-SVID.
```

Author: Fair, I created https://github.com/oauth-wg/oauth-spiffe-client-authentication/issues/46 for it.

To authenticate using an X.509-SVID, the client establishes a mutual TLS connection with the authorization server using its X.509-SVID as the client certificate. The authorization server validates the client certificate as an X.509-SVID and extracts the SPIFFE ID from the URI SAN. The server certificate MUST be validated by the client using its system trust store, and NOT the SPIFFE trust bundle.

> The request MUST include the `client_id` parameter identifying the client. The `client_id` MAY be the SPIFFE ID of the client, but it MAY also be any other client identifier recognized by the authorization server, such as a URL pointing to a Client ID Metadata Document ({{I-D.ietf-oauth-client-id-metadata-document}}). The authorization server determines the relationship between the `client_id` and the SPIFFE ID carried in the URI SAN of the presented X509-SVID as part of client authentication.

Author: Created https://github.com/oauth-wg/oauth-spiffe-client-authentication/issues/45 for this.

The server validates the client certificates according the following rules

1. Perform standard X.509 path validation against the trust anchors according to {{spiffe-bundle-validation}}.
2. Verify that the certificate contains exactly one URI SAN with a valid SPIFFE ID.
3. Verify that the certificate is a leaf certificate (Basic Constraints extension has CA=FALSE).
4. Verify that the certificate has the `digitalSignature` key usage bit set.

> 5. Verify that the SPIFFE ID in the URI SAN matches or is associated with a recognized client identifier. If the client authentication is presented with a `client_id` that is a URL described in {{I-D.ietf-oauth-client-id-metadata-document}}, verify that the SPIFFE ID in the URI SAN matches the `spiffe_id` value in the Client ID Metadata Document as described in {{client-registration-metadata}}.

Author: Copilot has a point here and I believe it also applies to JWT-SVID too ..

> 5. Verify that the SPIFFE ID in the URI SAN matches or is associated with a recognized client identifier. If the client authentication is presented with a `client_id` that is a URL described in {{I-D.ietf-oauth-client-id-metadata-document}}, verify that the SPIFFE ID in the URI SAN matches the `spiffe_id` value in the Client ID Metadata Document as described in {{client-registration-metadata}}.

Author: Fixed with [8a07edf](https://github.com/oauth-wg/oauth-spiffe-client-authentication/pull/44/commits/8a07edfe66863fb6961b74ff831e4fd47826537e)

To authenticate using an X.509-SVID, the client establishes a mutual TLS connection with the authorization server using its X.509-SVID as the client certificate. The authorization server validates the client certificate as an X.509-SVID and extracts the SPIFFE ID from the URI SAN. The server certificate MUST be validated by the client using its system trust store, and NOT the SPIFFE trust bundle.

> The request MUST include the `client_id` parameter identifying the client. The `client_id` MAY be the SPIFFE ID of the client, but it MAY also be any other client identifier recognized by the authorization server, such as a URL pointing to a Client ID Metadata Document ({{I-D.ietf-oauth-client-id-metadata-document}}). In either case, the SPIFFE ID carried in the URI SAN of the presented X509-SVID MUST be authorized to authenticate as the client identified by the `client_id`, as described below.

Reviewer: Editorial suggestion

```suggestion
The request MUST include the `client_id` parameter identifying the client. The `client_id` MAY be the SPIFFE ID of the client, or other client identifier recognized by the authorization server, such as a URL pointing to a Client ID Metadata Document ({{I-D.ietf-oauth-client-id-metadata-document}}). In either case, the SPIFFE ID carried in the URI SAN of the presented X509-SVID MUST be authorized to authenticate as the client identified by the `client_id`, as described below.
```
